### DISCUSSION TITLE
Have number types extend from a common type

### DISCUSSION THREAD

**Discussion by frankhaugen** [dotnet/csharplang#4572] [Language Ideas]
Applying generic constraints for "numeric" values are impossible, so I would propose that int, decimal, short, (and all the other numeric types), is implementing a common interface, something like "INumericValueBase", that doesn't have to do anything except allowing for a generic constraint, (Something<T>(T value) where T : INumericValueBase)

I know Mads have some very complicated proposal for something that will give this, but I prefer KISS-principled solutions

**Comment by yaakov-h** [comment]
If this was feasible without shapes/extensions, it would have been done already.

**Reply by bernd5** [reply]
It is feasable to add an Interface to all numeric types - but what can you do with such an interface?
Just as marker?
In generic algorithm you typically want to do something with the value / values - like adding...

**Comment by HaloFour** [comment]
The runtime team experimented with this a while ago, by adding an `IArithmetic` interface to the numeric types.  There are still some [vestiges of it in the source](https://referencesource.microsoft.com/#mscorlib/system/int32.cs,33).  The problem was that the performance was terrible compared to native arithmetic operations and that the semantics were painful since you needed to invoke instance members instead of being able to use operators.  The language could've theoretically smoothed over the latter point, but the performance issues would've still been there.
